File: game-server/src/main/java/net/dodian/jobs/impl/ShopProcessor.java
package net.dodian.jobs.impl;

import net.dodian.uber.game.Constants;
import net.dodian.uber.game.Server;
import net.dodian.uber.game.model.ShopHandler;
import net.dodian.uber.game.model.entity.player.Client;
import net.dodian.uber.game.model.entity.player.PlayerHandler;
import org.quartz.DisallowConcurrentExecution;
import org.quartz.Job;
import org.quartz.JobExecutionContext;
import org.quartz.JobExecutionException;

@DisallowConcurrentExecution

public class ShopProcessor implements Job {
    public void execute(JobExecutionContext context) throws JobExecutionException {
        boolean DidUpdate = false;
        for (int i = 1; i <= ShopHandler.MaxShops; i++) {
            if (ShopHandler.ShopItemsDelay[i] >= ShopHandler.MaxShowDelay) {
                for (int j = 0; j < ShopHandler.MaxShopItems; j++) {
                    if (ShopHandler.ShopItems[i][j] > 0) {
                        if (j < ShopHandler.ShopItemsStandard[i]
                                && ShopHandler.ShopItemsN[i][j] != ShopHandler.ShopItemsSN[i][j]) {
                            if (ShopHandler.ShopItemsN[i][j] < ShopHandler.ShopItemsSN[i][j]) {
                                double restockAmount = (ShopHandler.ShopItemsSN[i][j] - ShopHandler.ShopItemsN[i][j]) * 0.05;
                                ShopHandler.ShopItemsN[i][j] += restockAmount > 1 ? (int) restockAmount : 1;
                            } else Server.shopHandler.DiscountItem(i, j);
                        }
                        if (j >= ShopHandler.ShopItemsStandard[i]) Server.shopHandler.DiscountItem(i, j);
                        DidUpdate = true;
                    }
                }
            } else ShopHandler.ShopItemsDelay[i]++;
            if (DidUpdate) {
                for (int k = 1; k < Constants.maxPlayers; k++) {
                    if (PlayerHandler.players[k] != null) {
                        if (PlayerHandler.players[k].isShopping() && PlayerHandler.players[k].MyShopID == i) {
                            ((Client) PlayerHandler.players[k]).checkItemUpdate();
                        }
                    }
                }
                ShopHandler.ShopItemsDelay[i] = 0;
                DidUpdate = false;
            }
        }

    }

}

File: game-server/src/main/kotlin/net/dodian/uber/game/engine/processing/ShopProcessor.kt
package net.dodian.uber.game.engine.processing

import net.dodian.uber.game.Server
import net.dodian.uber.game.shop.ShopManager
import net.dodian.uber.game.model.entity.player.Client
import net.dodian.uber.game.engine.systems.world.player.PlayerRegistry

class ShopProcessor : Runnable {
    override fun run() {
        var didUpdate = false
        for (i in 0 until ShopManager.MaxShops) {
            if (ShopManager.ShopItemsDelay[i] >= ShopManager.MaxShowDelay) {
                for (j in 0 until ShopManager.MaxShopItems) {
                    if (ShopManager.ShopItems[i][j] > 0) {
                        if (j < ShopManager.ShopItemsStandard[i] && ShopManager.ShopItemsN[i][j] != ShopManager.ShopItemsSN[i][j]) {
                            if (ShopManager.ShopItemsN[i][j] < ShopManager.ShopItemsSN[i][j]) {
                                val restockAmount = (ShopManager.ShopItemsSN[i][j] - ShopManager.ShopItemsN[i][j]) * 0.05
                                ShopManager.ShopItemsN[i][j] += if (restockAmount > 1) restockAmount.toInt() else 1
                            } else {
                                Server.shopManager.DiscountItem(i, j)
                            }
                        }
                        if (j >= ShopManager.ShopItemsStandard[i]) {
                            Server.shopManager.DiscountItem(i, j)
                        }
                        didUpdate = true
                    }
                }
            } else {
                ShopManager.ShopItemsDelay[i]++
            }

            if (didUpdate) {
                PlayerRegistry.forEachActivePlayer { player ->
                    if (player.isShopping && player.MyShopID == i) {
                        (player as Client).checkItemUpdate()
                    }
                }
                ShopManager.ShopItemsDelay[i] = 0
                didUpdate = false
            }
        }
    }
}
